src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -735,7 +737,27 @@ DatabaseShardImp::stop()
     // Wait for the node store import thread
     // if necessary
     if (databaseImporter_.joinable())
+    {
+        // Tells the import function to halt
+        haltDatabaseImport_ = true;
+
+        // Wait for the function to exit
+        while (databaseImportStatus_)
+        {
+            // Unlock just in case the import
+            // function is waiting on the mutex
+            lock.unlock();
+
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            lock.lock();
+        }
+
+        // Calling join while holding the mutex_ without
+        // first making sure that doImportDatabase has
+        // exited could lead to deadlock via the mutex
+        // acquisition that occurs in that function
         databaseImporter_.join();
```

Comment:
Because the lock was released and reacquired, we can't be sure that `databaseImporter_` is still `joinable()` here. This is particularly true if you make the change I suggested elsewhere to do `databaseImporter_.detach()` at the end of the thread.
```
if (databaseImporter_.joinable())
    databaseImporter_.join();
```

Reply:
Done